solver/src/analytics.rs

```diff
@@ -0,0 +1,159 @@
+use crate::settlement::Settlement;
+use bigdecimal::Zero;
+use model::order::OrderUid;
+use num::BigRational;
+use shared::conversions::U256Ext;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+/// Record metric with surplus achieved in winning settlement
+/// vs that which was unrealized in other feasible solutions.
+pub fn num_unsettled_orders_with_better_surplus(
+    submitted: &Settlement,
+    all: impl Iterator<Item = Settlement>,
+) -> usize {
+    let submitted_prices = submitted
+        .clearing_prices()
+        .iter()
+        .map(|(token, price)| (token, price.to_big_rational()))
+        .collect::<HashMap<_, _>>();
+    let submitted_surplus: HashMap<_, _> = submitted
+        .trades()
+        .iter()
+        .map(|trade| {
+            (
+                trade.order.order_meta_data.uid,
+                trade
+                    .surplus(
+                        &submitted_prices[&trade.order.order_creation.sell_token],
+                        &submitted_prices[&trade.order.order_creation.buy_token],
+                    )
+                    .unwrap_or_else(BigRational::zero),
+            )
+        })
+        .collect();
+
+    let mut best_surplus: HashMap<OrderUid, BigRational> = HashMap::new();
+    for settlement in all {
+        let trades = settlement.trades();
+        let clearing_prices = settlement
+            .clearing_prices()
+            .iter()
+            .map(|(token, price)| (token, price.to_big_rational()))
+            .collect::<HashMap<_, _>>();
+
+        for trade in trades {
+            let order_id = trade.order.order_meta_data.uid;
+            if submitted_surplus.contains_key(&order_id) {
+                let surplus = trade
+                    .surplus(
+                        &clearing_prices[&trade.order.order_creation.sell_token],
+                        &clearing_prices[&trade.order.order_creation.buy_token],
+                    )
+                    .unwrap_or_else(BigRational::zero);
+                let entry = best_surplus.entry(order_id);
+                match entry {
+                    Entry::Occupied(mut entry) => {
+                        let value = entry.get_mut();
+                        *value = (value.clone()).max(surplus);
+                    }
+                    Entry::Vacant(entry) => {
+                        entry.insert(surplus);
+                    }
+                }
+            }
+        }
+    }
+
+    let mut count = 0;
+    for (order_id, submitted) in submitted_surplus.iter() {
+        let best = best_surplus.get(order_id).expect("exists by construction");
+        if best > &(BigRational::new(101.into(), 100.into()) * submitted) {
+            count += 1;
+            tracing::warn!("Found Matched order {:?} whose submission surplus was 1% worse than in another, valid, lower ranked settlement", order_id)
+        }
+    }
+    count
```

Comment:
This is a great suggestion. I will have to tinker a bit with the definition of buckets since there is a division by zero singularity when the submitted surplus is zero.

Reply:
Division by 0 shouldn't be possible if we do surplus difference (i.e. `settled_trade.surplus() - solved_but_not_settled_trade.surplus()`).

Reply:
Yes but that difference alone is not a relative measure and won't fit well into buckets without some sort of normalization. Normalization of this difference is where the zero division comes from.

Reply:
Sorry, by `.surplus()` I meant `.surplus_percentage()`. Technically, division by 0 can happen only if the buy/sell amounts are 0 right? Since orders with 0 amounts aren't allowed by the API[1] it shouldn't ever happen. Either way, we can just skip orders with 0 amounts to avoid panics if the orderbook every misbehaves.

[1]: https://github.com/gnosis/gp-v2-services/blob/d52387e96d7a0ffbae62b547f0b79ad88460c2f5/orderbook/src/api/order_validation.rs#L297-L299